### PR TITLE
Fix promise assertions

### PR DESCRIPTION
**Pull request by ariporad**
Hi All,

Here's the fix for #309, It has a good number of tests, but let me know if there's anything I should change.

### CODE REVIEW

File: lib/test.js [hunk 1]
```diff
@@ -165,35 +172,49 @@ Test.prototype._end = function (err) {
 Test.prototype.exit = function () {
 	var self = this;
 
-	// calculate total time spent in test
-	this.duration = globals.now() - this._timeStart;
+	function checkAssertLength() {
```
Comment (ariporad) [inline lib/test.js]: I'm not sure if this should be here, or it should be a method, or if it should be an external function which accepts `self`. Thoughts?

File: lib/test.js [hunk 2]
```diff
@@ -25,11 +25,18 @@ function Test(title, fn) {
 
 	this.title = title || fnName(fn) || '[anonymous]';
 	this.fn = isGeneratorFn(fn) ? co.wrap(fn) : fn;
-	this.assertCount = 0;
+	this.assertPromises = [];
```
Comment (jamestalmage) [inline lib/test.js]: `assertPromises` -> `assertions`

File: lib/test.js [hunk 3]
```diff
@@ -230,33 +251,23 @@ Test.prototype._publicApi = function () {
 	api.plan = this.plan.bind(this);
 
 	function skipFn() {
-		self._assert();
+		self._assert(Promise.resolve());
```
Comment (jamestalmage) [inline lib/test.js]: `self._assert()` should work fine here. `Promise.all` converts array members to promise using `Promise.resolve` anyways.

File: test/test.js [hunk 4]
```diff
@@ -2,6 +2,14 @@
 var test = require('tap').test;
 var _ava = require('../lib/test');
 
+function delay(val, time) {
```
Comment (jamestalmage) [inline test/test.js]: use [`delay`](https://www.npmjs.com/package/delay)

File: test/test.js [hunk 5]
```diff
@@ -351,3 +359,119 @@ test('skipped assertions count towards the plan', function (t) {
 		t.end();
 	});
 });
+
+test('throws and doesNotThrow work with promises', function (t) {
```
Comment (jamestalmage) [inline test/test.js]: make two separate tests.

File: test/test.js [hunk 6]
```diff
@@ -351,3 +359,119 @@ test('skipped assertions count towards the plan', function (t) {
...
+	}).run().then(function (a) {
+		t.ifError(a.assertError);
+		t.is(a.planCount, 2);
+		t.is(a.assertCount, 2);
+		t.is(asyncCalled, true);
+		t.end();
+	});
+});
+
+test('t.end is called when a promise passed to t.throws hasn\'t resolved yet', function (t) {
```
Comment (jamestalmage) [inline test/test.js]: this makes it sound like AVA will call `t.end` itself. I suggest something like:

`waits for t.throws to resolve after t.end is called`

This comment applies to multiple tests.

File: test/test.js [hunk 7]
```diff
@@ -351,3 +359,119 @@ test('skipped assertions count towards the plan', function (t) {
...
+		}, 5);
+	}).run().catch(function (err) {
+		t.is(err.operator, 'plan');
+		t.is(err.actual, 3);
+		t.is(err.expected, 2);
+		t.end();
+	});
+});
+
+test('number of assertions doesn\'t t.plan when the test exits, but before all promises resolve another is added', function (t) {
```
Comment (jamestalmage) [inline test/test.js]: > number of assertions doesn't **match** plan

missing a word, and (minor nit) I think it reads better without the `t.`

File: test/test.js [hunk 8]
```diff
@@ -2,6 +2,14 @@
 var test = require('tap').test;
 var _ava = require('../lib/test');
```
Comment (jamestalmage) [inline test/test.js]:
```js
var Promise = require('bluebird');
```

Promise is not globally available in Node `0.10`

File: lib/test.js [hunk 9]
```diff
@@ -165,35 +172,49 @@ Test.prototype._end = function (err) {
...
+		if (self.assertError === undefined && self.planCount !== null && self.planCount !== self.assertPromises.length) {
+			self._setAssertError(new assert.AssertionError({
+				actual: self.assertPromises.length,
+				expected: self.planCount,
+				message: 'Assertion count does not match planned',
+				operator: 'plan'
+			}));
+
+			self.assertError.stack = self.planStack;
+			return false;
```
Comment (jamestalmage) [inline lib/test.js]: doesn't look like you use these return values.

**Comment by jamestalmage** [comment]
@ariporad - Good work! A few minor nits, but overall very well done!

File: lib/test.js (hunk 3 above)
Comment (ariporad) [inline lib/test.js]: I think we should keep that here, as it makes it more obvious what we're doing, and it makes it a ***lot*** easier to reason about `this.assertions`.

Comment (jamestalmage) [inline lib/test.js]: Agreed. I wrote this before I saw how you optimized the `onAssertion` callback. You can disregard.

**Comment by ariporad** [comment]
@jamestalmage: Should be fixed now.

**Comment by ariporad** [comment]
@jamestalmage: Hang on a second actually, delay doesn't like node@<4.0.0

**Comment by sindresorhus** [comment]
> delay doesn't like node@<4.0.0

It's `node@<0.12`, and just put `global.Promise = require('bluebird');` at the top.

**Comment by ariporad** [comment]
@sindresorhus: Yeah, I was trying to avoid that, but it's not that big a deal.

File: package.json [hunk 10]
```diff
@@ -93,6 +93,7 @@
     "core-assert": "^0.1.0",
     "debug": "^2.2.0",
     "deeper": "^2.1.0",
+    "delay": "^1.2.0",
```
Comment (sindresorhus) [inline package.json]: This should be a dev dependency.

Comment (ariporad) [inline package.json]: Fixed.

File: lib/test.js (hunk 1 above)
Comment (sindresorhus) [inline lib/test.js]: I would make it a private instance method (prefixed with `_`). The function checks if assertion count matches plan count. I think the name could be clearer.

Comment (ariporad) [inline lib/test.js]: Done.

File: lib/test.js [hunk 11]
```diff
@@ -181,19 +180,38 @@ Test.prototype.exit = function () {
 
 		this.assertError.stack = this.planStack;
 	}
+};
 
-	if (!this.ended) {
-		this.ended = true;
+Test.prototype.exit = function () {
```
Comment (sindresorhus) [inline lib/test.js]: remove this empty line

Comment (ariporad) [inline lib/test.js]: @sindresorhus: Fixed

**Comment by ariporad** [comment]
Ok, @jamestalmage, @sindresorhus: Everything should be fixed now.

**Comment by vdemedes** [comment]
LGTM, great stuff!

**Comment by jamestalmage** [comment]
Let's get https://github.com/sindresorhus/delay/pull/8 merged and published. Then use the updated function here. Once that happens, I think we are good to go.

**Comment by ariporad** [comment]
Ok, I just rebased onto master.

**Comment by sindresorhus** [comment]
@ariporad delay@1.3.0 is now out. Can you update tests to use it? Other than that, this looks good to land.

**Comment by ariporad** [comment]
@sindresorhus: Ok. Fixed, and rebased onto master. Ready for merging.

**Comment by sindresorhus** [comment]
Landed. Really dig this change @ariporad :)

**Pull request closed without merging**